Incorporación de markup_doc, model_ai y utilidades en core

eduranm opened the pull request:
Creación y ampliación de la app core (modelos comunes, formularios, utilidades, vistas y hooks).

Inclusión de app markup_doc para el procesamiento y marcado de artículos DOCX (modelos, API REST, tareas y hooks).

Inclusión de la librería markuplib con funciones para DOCX y transformaciones OMML → MathML.

Creación de la app model_ai para integración con modelos LLaMA, utilidades de inferencia y tareas de IA.

Módulos en reference (configuración de Gemini y utilidades de datos).

Consolidación y limpieza de migraciones en reference, core y core_settings.

eduranm committed:
Integración de ArticleViewSet en el enrutador y actualización de URLs principales

eduranm committed:
Actualización de settings: cambio en directorio de modelos, inclusión de nuevas apps y aumento del límite de campos

eduranm committed:
Ampliación de core.models: nuevos modelos de género, idioma, licencias, textos con idioma y manejo flexible de fechas

eduranm committed:
Actualización en Reference: uso de ReferenceStatus y ajuste en creación de referencias

eduranm committed:
Refactor en reference.config: serialización de ejemplos con json.dumps y ampliación de tipos soportados (confproc, full_text, etc.)

eduranm committed:
Refactor en marker: actualización de importación de GenericLlama desde model_ai

eduranm committed:
Eliminación de reference.tasks (responsabilidad movida a data_utils y model_ai)

eduranm committed:
Refactor en wagtail_hooks: uso de SnippetViewSet y soporte para descarga de modelos

eduranm committed:
Actualización de dependencias: inclusión de langid, google-generativeai y python-docx

eduranm committed:
Creación y ampliación de la app core: modelos comunes, formularios, vistas de búsqueda, utilidades y hooks de Wagtail

eduranm committed:
Nueva app markup_doc: modelos ArticleDocx, API REST, tareas de marcado, utilidades y hooks de Wagtail

eduranm committed:
Nueva librería markuplib: funciones para procesar DOCX y transformaciones OMML a MathML

eduranm committed:
Nueva app model_ai: integración de LLaMA, funciones genéricas, mensajes de inferencia, tareas y hooks de Wagtail

eduranm committed:
Módulos adicionales en reference: configuración de Gemini y utilidades de procesamiento de datos

eduranm committed:
Consolidación de migraciones en reference: actualización de 0001_initial.py y eliminación de migraciones intermedias

eduranm committed:
Actualización de migración inicial en core_settings: cambio de versión de Django y traducción de verbose_name a inglés

eduranm committed:
Creación de migración inicial en core

eduranm committed:
Eliminación de la app llama3: funcionalidades migradas a model_ai

robertatakenaka commented:
@eduranm he contratado el servicio copilot. Por favor, evalue las recomendaciones de correcciones si se aplican. Cambia en tu código local y después suba como nuevos commits.

eduranm committed:
fix(markup_doc): capturar ArticleDocxMarkup.DoesNotExist en generate_xml

Corrige el tipo de excepción para responder 404 cuando el registro no existe.

eduranm committed:
refactor(markuplib): eliminar prints de depuración en extracción de enlaces

Reduce ruido en logs y mantiene la función enfocada a su retorno.

eduranm committed:
style(model_ai): reemplazar bare except por Exception en download_model

Mejora legibilidad y buenas prácticas de manejo de errores.

eduranm committed:
fix(reference): corregir literales y separadores en config_gemini para prompt de referencias

Se agregan comillas a campos textuales y se corrigen comas/keys para evitar errores de parseo del prompt.

eduranm committed:
i18n(reference): envolver labels y help_text con gettext_lazy

Permite traducción de 'Mixed Citation' y 'Rating from 1 to 10'.

robertatakenaka commented on core/models.py:
```diff
@@ -43,4 +52,487 @@ class CommonControlField(models.Model):
+        if mission:
+            return mission
+        
+        language_order = ['pt', 'es', 'en']
```
@eduranm verificar que hay otros tipos de idiomas

robertatakenaka commented on markup_doc/labeling_utils.py:
```diff
@@ -0,0 +1,1125 @@
+                    if ' e ' in surname_buscado:
+                        surname1 = surname_buscado.split(' e ')[0].strip().lower()
+                        surname2 = surname_buscado.split(' e ')[1].strip().lower()
+
```
```
# Código original con problemas identificados:
"""
PROBLEMAS DETECTADOS:
1. Código repetitivo - viola el principio DRY (Don't Repeat Yourself)
2. Las variables surname1 y surname2 se sobrescriben múltiples veces si hay varios separadores
3. No maneja casos donde el split() podría fallar (ej: separador al final)
4. No valida que existan exactamente 2 apellidos después del split
5. Lógica difícil de mantener y extender
"""

# VERSIÓN MEJORADA 1: Usando diccionario y break
def process_surnames_v1(surname_buscados):
    """
    Procesa una lista de apellidos buscando separadores comunes.
    Retorna una lista de tuplas (surname1, surname2) o None si no hay separador.
    """
    # Definir separadores en orden de prioridad
    SEPARATORS = [' y ', ' and ', ' & ', ' e ']
    
    results = []
    for surname_buscado in surname_buscados:
        surname_pair = None
        
        # Buscar el primer separador que aplique
        for separator in SEPARATORS:
            if separator in surname_buscado:
                parts = surname_buscado.split(separator)
                if len(parts) == 2:  # Validar que tengamos exactamente 2 partes
                    surname1 = parts[0].strip().lower()
                    surname2 = parts[1].strip().lower()
                    surname_pair = (surname1, surname2)
                    break  # Usar el primer separador encontrado
        
        results.append(surname_pair)
    
    return results


# VERSIÓN MEJORADA 2: Usando expresiones regulares (más flexible)
import re

def process_surnames_v2(surname_buscados):
    """
    Procesa apellidos usando regex para múltiples separadores.
    Más eficiente y flexible que múltiples if statements.
    """
    # Patrón que captura los separadores comunes
    pattern = re.compile(r'\s+(y|and|&|e)\s+', re.IGNORECASE)
    
    results = []
    for surname_buscado in surname_buscados:
        match = pattern.split(surname_buscado)
        if len(match) >= 3:  # [apellido1, separador, apellido2, ...]
            surname1 = match[0].strip().lower()
            surname2 = match[2].strip().lower()
            results.append((surname1, surname2))
        else:
            results.append(None)
    
    return results


# VERSIÓN MEJORADA 3: Generador con manejo robusto de errores
def process_surnames_v3(surname_buscados):
    """
    Versión más robusta con validación y logging opcional.
    Usa yield para eficiencia de memoria con listas grandes.
    """
    SEPARATORS = {
        ' y ': 'spanish_and',
        ' and ': 'english_and', 
        ' & ': 'ampersand',
        ' e ': 'portuguese_and'
    }
    
    for surname_buscado in surname_buscados:
        if not surname_buscado:  # Validar entrada vacía
            yield None
            continue
            
        found = False
        for separator, sep_type in SEPARATORS.items():
            if separator in surname_buscado:
                parts = surname_buscado.split(separator)
                
                # Validación más estricta
                if len(parts) == 2 and all(p.strip() for p in parts):
                    surname1 = parts[0].strip().lower()
                    surname2 = parts[1].strip().lower()
                    yield {
                        'surname1': surname1,
                        'surname2': surname2,
                        'separator_type': sep_type,
                        'original': surname_buscado
                    }
                    found = True
                    break
        
        if not found:
            yield None


# EJEMPLO DE USO:
if __name__ == "__main__":
    # Datos de prueba
    test_surnames = [
        "García y López",
        "Smith and Jones",
        "Brown & Williams",
        "Silva e Santos",
        "Martinez",  # Sin separador
        " y Solo",    # Caso edge: separador al inicio
        "Multiple y and separators"  # Múltiples separadores
    ]
    
    print("=== Versión 1 (simple) ===")
    for result in process_surnames_v1(test_surnames):
        print(result)
    
    print("\n=== Versión 2 (regex) ===")
    for result in process_surnames_v2(test_surnames):
        print(result)
    
    print("\n=== Versión 3 (robusta) ===")
    for result in process_surnames_v3(test_surnames):
        if result:
            print(f"{result['original']} -> {result['surname1']}, {result['surname2']} [{result['separator_type']}]")
        else:
            print("No separator found")


# RECOMENDACIÓN FINAL:
"""
Para tu caso en SciELO, recomendaría la versión 3 porque:

1. Es más mantenible y extensible
2. Proporciona metadata útil (tipo de separador usado)
3. Maneja casos edge de forma robusta
4. Usa generadores para eficiencia con grandes volúmenes de datos
5. Facilita el debugging y logging
6. Se integra bien con pipelines de procesamiento de artículos

Si necesitas procesar nombres de autores de artículos científicos,
considera también normalizar acentos y caracteres especiales:

from unicodedata import normalize

def normalize_name(name):
    return normalize('NFKD', name).encode('ASCII', 'ignore').decode('utf-8')
"""
```

pitangainnovare commented on core/models.py:
```diff
@@ -43,4 +52,487 @@ class CommonControlField(models.Model):
+        license_version = None
+        license_language = None
+
+        url = url.lower()
+        url_parts = url.split("/")
+        if not url_parts:
+            return {}
```
@eduranm

¿No sería mejor usar `urllib.parse` para convertir la cadena URL?

```python
from urllib.parse import urlparse, parse_qs

url = "https://example.com/path/page?arg1=val1&arg2=val2#section"

# quebra URL em componentes
parts = urlparse(url)
print(parts.scheme)   # https
print(parts.netloc)   # example.com
print(parts.path)     # /path/page
print(parts.query)    # arg1=val1&arg2=val2
print(parts.fragment) # section

# parse de query string
query_params = parse_qs(parts.query)
print(query_params)  # {'arg1': ['val1'], 'arg2': ['val2']}
```

pitangainnovare commented on core/models.py:
```diff
@@ -43,4 +52,487 @@ class CommonControlField(models.Model):
+                    pass
+                break
+
+        return dict(
+            license_type=license_type,
+            license_version=license_version,
+            license_language=license_language,
```
@eduranm

El método devuelve esto:

```python
return dict(
    license_type=license_type,
    license_version=license_version,
    license_language=license_language,
)
```

¿No sería mejor darle otro nombre al método? Por ejemplo, `extract_license_data_from_url`?

pitangainnovare commented on reference/models.py:
```diff
@@ -10,11 +10,23 @@
 # Create your models here
 class Reference(CommonControlField, ClusterableModel):
-    mixed_citation = models.TextField(_("Mixed Citation"), null=False, blank=True)
+    mixed_citation = models.TextField(("Mixed Citation"), null=False, blank=True)
```
@eduranm 

Necesita incluir el `_` antes de `(`


`mixed_citation = models.TextField(_("Mixed Citation"), null=False, blank=True)`

pitangainnovare commented on reference/models.py:
```diff
@@ -10,11 +10,23 @@
+        choices=ReferenceStatus.choices,
+        blank=True,
+        default=ReferenceStatus.NO_REFERENCE
+    )
```
@eduranm 

Vamos a estandarizar los nombres de los campos al idioma inglés (status)

pitangainnovare commented on reference/models.py:
```diff
@@ -45,7 +53,7 @@ class ElementCitation(Orderable):
             MaxValueValidator(10)  # Máximo 10
         ],
-        help_text=_("Rating from 1 to 10")
+        help_text="Rating from 1 to 10"
```
@eduranm 

¿Hay alguna razón especial para eliminar la traducción `_()`?

pitangainnovare commented on reference/wagtail_hooks.py:
```diff
@@ -8,12 +8,22 @@
 )
 
+from wagtail.snippets.views.snippets import (
+    CreateView,
+    EditView,
```
@eduranm 

¿Vamos a usar `CreateView` a partir de `modeladmin` o de `snippets`?

pitangainnovare replied:
Sugiero borrar la línea 8

pitangainnovare commented on markup_doc/tasks.py:
```diff
@@ -0,0 +1,326 @@
+from model_ai.function_llama import functionsLlama
+from reference.config_gemini import create_prompt_reference
+
+def clean_labels(text):
+    # Eliminar etiquetas tipo [kwd] o [sectitle], incluso si tienen espacios como [/ doctitle ]
+    text = re.sub(r'\[\s*/?\s*\w+(?:\s+[^\]]+)?\s*\]', '', text)
+
+    # Reemplazar múltiples espacios por uno solo
+    text = re.sub(r'[ \t]+', ' ', text)
+
+    # Eliminar espacios antes de los signos de puntuación
+    text = re.sub(r'\s+([;:,.])', r'\1', text)
+
+    # Normalizar múltiples saltos de línea
+    text = re.sub(r'\n+', '\n', text)
+
+    # Quitar espacios al principio y final
+    return text.strip()
```
Sería mejor colocar este código en algún módulo `utils`.

pitangainnovare commented on markup_doc/tasks.py:
```diff
@@ -0,0 +1,326 @@
+    instance = MarkupXML.objects.get(id=instance_id)
+    content_head = instance_content
+    content_body_dict = instance_content_body
+    xml, stream_data_body = get_xml(instance, content_head, content_body_dict, instance_content_back)
```
¿No sería interesante incluir algún manejo con `try/except` aqui?

pitangainnovare commented on markup_doc/tasks.py:
```diff
@@ -0,0 +1,326 @@
+
+
+@celery_app.task()
+def get_labels(title, user_id):
```
Este método es muy grande. Después vamos a evaluar cómo dividirlo en varias partes más pequeñas.

pitangainnovare commented on markuplib/function_docx.py:
```diff
@@ -0,0 +1,589 @@
+ImageModel = get_image_model()
+
+
+class functionsDocx:
```
Es una convención escribir el nombre de las clases comenzando las letras con mayúscula (FunctionsDocx). Sugiero otro nombre, algo como DocxUtil.

pitangainnovare commented on markuplib/function_docx.py:
```diff
@@ -0,0 +1,589 @@
+
+class functionsDocx:
+
+    def openDocx(filename):
```
En el caso de los métodos, es costumbre usar todo en minúsculas, separando palabras con guion bajo (`snake_case`).

pitangainnovare replied:
open_docx

eduranm committed:
model_ai: corrige etiqueta de i18n y refuerza unicidad de LlamaModel en save()

pitangainnovare commented on markuplib/function_docx.py:
```diff
@@ -0,0 +1,589 @@
+        namespaces = '{http://schemas.openxmlformats.org/wordprocessingml/2006/main}'
+
+        # Abrir el archivo DOCX como un archivo ZIP
+        with zipfile.ZipFile(docx_path, 'r') as docx:
```
@eduranm 

Sugiero que haya un método que abra el DOCX en modo zip y lo guarde en modo zip, y aquí los uses.

Ve una versión que tenemos en:

```python
def unzip_docx(path, prefix="scl_xml2pdf"):
    """
    Extrae el contenido de un archivo DOCX en un directorio temporal.
    Crea un nuevo directorio temporal con el prefijo especificado, extrae todos los archivos del archivo DOCX en él
    y devuelve la ruta al directorio temporal que contiene los archivos extraídos.

    Args:
        path (str): La ruta al archivo DOCX a extraer.
        prefix (str): El prefijo a usar para el directorio temporal.

    Returns:
        str: La ruta al directorio temporal que contiene los archivos extraídos.
    """
    temp_docx_source_dir = tempfile.mkdtemp(prefix=prefix)

    with zipfile.ZipFile(path, 'r') as zf:
        zf.extractall(temp_docx_source_dir)

    return temp_docx_source_dir

def embed_docx(source_dir, prefix="scl_xml2pdf", suffix=".docx"):
    """
    Crea un archivo DOCX comprimiendo todos los archivos del directorio fuente en un archivo ZIP con extensión .docx.
    Después de la compresión exitosa, elimina el directorio fuente y devuelve la ruta al archivo DOCX recién creado.

    Args:
        source_dir (str): La ruta al directorio que contiene los archivos a comprimir.
        prefix (str): El prefijo a usar para el archivo DOCX temporal.
        suffix (str): El sufijo a usar para el archivo DOCX temporal.

    Returns:
        str: La ruta al archivo DOCX recién creado.
    """
    temp_docx_path = tempfile.mktemp(prefix=prefix, suffix=suffix)
    with zipfile.ZipFile(temp_docx_path, 'w') as zf:
        for root, _, files in os.walk(source_dir):
            for f in files:
                fpath_absolute = os.path.join(root, f)
                fpath_rel = os.path.relpath(fpath_absolute, source_dir)
                zf.write(fpath_absolute, arcname=fpath_rel)
    
    try:
        shutil.rmtree(source_dir)
    except OSError:
        raise DirectoryRemovalError(f'No se pudo eliminar el directorio temporal: {source_dir}')
    return temp_docx_path
```

https://github.com/pitangainnovare/packtools/blob/1287257bfb0320c6082f365dc9d02e357c17fda8/packtools/sps/formats/pdf/utils/file_utils.py#L48

pitangainnovare commented on markuplib/function_docx.py:
```diff
@@ -0,0 +1,589 @@
+        return ' '.join(links) if links else None
+
+
+    def extractContent(self, doc, doc_path):
```
extract_content

pitangainnovare replied:
@eduranm 

Después vamos a necesitar reducir el tamaño de la mayoría de los métodos siguientes. Algo mucho más pequeño, cada método haciendo solo una cosa.

eduranm committed:
reference: renombra campo 'estatus' a 'status' y ajusta etiqueta a 'Reference status' (incluye migraciones)

pitangainnovare commented on markuplib/function_docx.py:
```diff
@@ -0,0 +1,589 @@
+
+            return clean_text.strip()
+
+        def extrae_Tabla(element, rels_map, namespaces):
```
extrae_table (o extract_table)

eduranm committed:
reference: elimina import no utilizado en wagtail_hooks.py

pitangainnovare commented on model_ai/generic_llama.py:
```diff
@@ -0,0 +1,35 @@
+from config.settings.base import LLAMA_MODEL_DIR
+from llama_cpp import Llama
+import os
+import google.generativeai as genai
+from model_ai.models import LlamaModel
+
+class GenericLlama:
+
+  def __init__(self, messages=None, response_format=None, max_tokens=4000, temperature=0.1, top_p=0.1, type='chat'): 
+    model_ai = LlamaModel.objects.first()
+    self.llm = Llama(model_path = os.path.join(LLAMA_MODEL_DIR, model_ai.name_file), n_threads=2, n_ctx=5000)
+    self.messages = messages
+    self.response_format = response_format
+    self.max_tokens = max_tokens
+    self.temperature = temperature
+    self.top_p = top_p
+    self.type = type
+
+  def run(self, user_input):
+    if self.type == 'chat':
+      input = self.messages.copy()
+      input.append({
+        'role': 'user',
+        'content': user_input
+      })
+      return self.llm.create_chat_completion(messages=input, response_format=self.response_format, max_tokens=self.max_tokens, temperature=self.temperature, top_p=self.top_p)
+    if self.type == 'prompt':
+      #USO DE GEMINI
+      model_ai = LlamaModel.objects.first()
+      if model_ai and model_ai.api_key_gemini:
+        genai.configure(api_key=model_ai.api_key_gemini)
+        model = genai.GenerativeModel('models/gemini-2.0-flash')
+        return model.generate_content(user_input).text
+      else:
+        return self.llm(user_input, max_tokens=2000, temperature=self.temperature, stop=["\n\n"])
```
@eduranm 

Aquí hay una versión actualizada del `GenericLlama`, pensando en el manejo de errores y en la optimización usando Singleton: [https://github.com/pitangainnovare/markapi/blob/main/llama3/generic_llama.py](https://github.com/pitangainnovare/markapi/blob/main/llama3/generic_llama.py)

pitangainnovare commented on model_ai/models.py:
```diff
@@ -0,0 +1,75 @@
+        self.attrs.update({'disabled': 'disabled'})
+
+
+class LlamaModel(CommonControlField):
```
LLMModel ?

pitangainnovare commented on model_ai/wagtail_hooks.py:
```diff
@@ -0,0 +1,119 @@
+from model_ai.tasks import download_model
+
+
+class LlamaModelCreateView(CreateView):
```
Llama es solo uno de los muchos posibles modelos. ¿Qué tal si llamamos al modelo `LLMModel` o `AIModel`?

LLMModelCreateView

pitangainnovare reviewed (changes requested):
@eduranm 
@robertatakenaka 

Hay algunos métodos muy grandes que podrían reducirse a métodos pequeños. Esto facilitaría el mantenimiento del código, sobre todo en la parte que manipula datos en Docx. Llegué a implementar alguna forma de instanciar el `GenericLlama` usando el patrón Singleton y con algo de manejo de errores. Podría ser interesante aprovechar ese código en el `GenericLlama` de este PR.
